### PR TITLE
TableDefinition refactoring: holding full schema information, derived…

### PR DESCRIPTION
**Pull request by arkadius**
… types are determined lazy based on this

## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

**Review by mslabek — APPROVED** [review]

File: engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/source/FlinkMiniClusterTableOperations.scala [hunk 1]
```diff
@@ -25,6 +26,7 @@ object FlinkMiniClusterTableOperations extends LazyLogging {
     ThreadUtils.withThisAsContextClassLoader(getClass.getClassLoader) {
       implicit val env: StreamTableEnvironment = MiniClusterEnvBuilder.buildStreamTableEnv
       val (inputTablePath, inputTableName)     = createTempFileTable(schema)
+      logger.info(s"Parser schem: $schema")
```
Comment (mslabek) [inline engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/source/FlinkMiniClusterTableOperations.scala:29]: leftover log

File: engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/sink/TableSink.scala [hunk 2]
```diff
@@ -61,7 +60,9 @@ class TableSink(
 
     val inputValueTable = tableEnv
       .fromDataStream(streamOfRows)
-      .select(tableDefinition.sinkRowDataType.getLogicalType.toRowTypeUnsafe.getFieldNames.asScala.toList.map($): _*)
+      .select(
+        tableDefinition.schema.toSinkRowDataType.getLogicalType.toRowTypeUnsafe.getFieldNames.asScala.toList.map($): _*
```
Comment (mslabek) [inline engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/sink/TableSink.scala:64]: 1. Doesn't it have to be sorted like in table source?
2. Thats a pretty long expression and its used in similar forms in other places. Maybe its worth to extract some of it to the `TableDefinition` to increase readability and deduplicate some code?

**Review by mslabek — APPROVED** [review]

File: engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/source/TableSource.scala [hunk 3]
```diff
@@ -89,26 +93,54 @@ class TableSource(
...
         )
     }
   }
 
+  // This is copy-paste of toSourceRowDataType with filtering-out of computed columns.
+  // We dont' want to generate data for computed columns - they will be added during parsing of test data
+  private def fieldsWithoutComputedColumns =
+    tableDefinition.schema.getColumns.asScala
+      .filterNot(c => c.isInstanceOf[ComputedColumn])
+      .map(c => DataTypes.FIELD(c.getName, DataTypeUtils.removeTimeAttribute(c.getDataType)))
```
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/source/TableSource.scala:141]: I want to be as compatible with `toSourceRowDataType` as it is possible - In the comment above there is a notice that it is a copy-pasted method, I can add also a comment next to this line. TBH I don't know why they remove these attributes.

[the remark this answers is not in the document]
Comment (arkadius) [reply engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/source/TableSource.scala]: I extracted an extension class for `ResolvedSchema` to make it clear that it is a copy-paste. I also copied their explanation why they remove this attribute

File: engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/sink/TableSink.scala (hunk 2 above)
Comment (arkadius) [reply engine/flink/components/table/src/main/scala/pl/touk/nussknacker/engine/flink/table/sink/TableSink.scala]: 1. No, it doesn't. We should insert columns in the same order as in sink. This sorting is only needed when a Row will be serialized using TestInfo prepared by our detection
2. I extracted fields in `TableDefinition` and added some extensions to make these lines shorter

**Pull request merged**